File: python_solution/problem16.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Aug 27 20:34:09 2024

@author: sharminalina
"""
import sys
class Problem16:
    def threeSumClosest(self, nums: list[int],target)->int:
        closestSum = sys.maxsize
        n = len(nums)
                

        if n < 3:
            closestSum = 0
        else:
            new_nums = sorted(nums)
            for i in range(n-2):
                low = i+1
                high = n-1
                while(low < high):
                    current_sum = new_nums[i] + new_nums[low] + new_nums[high]
                    if (abs(closestSum -target) > abs(current_sum - target)):
                        closestSum = current_sum
                        low += 1
                        high -= 1
                    if current_sum > target:
                        high -= 1
                    elif current_sum < target:
                        low +=1
                    
        return closestSum


    
def main():
    # Problem 16: 3 Sum Closest
    nums_input = input("Enter the list of integers (comma-separated): ")
    # Convert input string to a list of integers
    nums = list(map(int, nums_input.split(',')))
    
    target = int(input("Enter a target number: "))
    
    solution_prob16 = Problem16()
    result = solution_prob16.threeSumClosest(nums,target)
    
    # Print the list of int whose sum is 0
    print(f"Closest sum of target : {result}")

if __name__ == "__main__":
    main()

File: C#Solutions/problem16.cs
using System;

namespace AllProblems
{
    public static class Problem16
    {
        public static void Solve()
        {
            int[] nums = [-1,2,1,-4];
            int target = 1;
            int result = CloseestSum(nums, target);
            Console.WriteLine("Closest sum: " + result);
        }

        public static int CloseestSum(int[] nums, int target)
        {
            int n = nums.Length;
            int closestSum = int.MaxValue;
            for(int i = 0; i < n-2; i++){
                int low = i+1;
                int high = n-1;

                while(low < high){
                    int currSum = nums[i]+nums[low] + nums[high];
                    if(Math.Abs(currSum -target) < Math.Abs(closestSum - target)){
                        closestSum = currSum;
                    }
                    if(currSum > target) low ++;
                    else high --;

                }

            }
            return closestSum;


            
        }

        
    }
}
